### PR TITLE
core: Fix race conditions, mostly around diffs

### PR DESCRIPTION
**Pull request by jbardin**
Add a mutex and some locked getters and setters to InstanceDiff, so that we can safely use them concurrently. It's not optimal that we expose these in the exported API, but they need to be accessed from the command and the schema packages too. 

Add a mutex to the MockHook, and lock all methods.

### CODE REVIEW

File: config/raw_config.go [hunk 1]
```diff
@@ -93,6 +93,8 @@ func (r *RawConfig) Value() interface{} {
 // structure will always successfully decode into its ultimate
 // structure using something like mapstructure.
 func (r *RawConfig) Config() map[string]interface{} {
+	r.lock.Lock()
+	defer r.lock.Unlock()
```
Comment (phinze) [inline config/raw_config.go]: This one is sort of a cheat since we're returning the pointer to the map anyways, yeah?

Comment (phinze) [inline config/raw_config.go]: Relevant questions:

 * Do callers require write access to the result returned?
 * If yes, should we do a copy here?

File: terraform/diff.go [hunk 2]
```diff
@@ -366,6 +375,17 @@ func (d *InstanceDiff) RequiresNew() bool {
 		return false
 	}
 
+	d.mu.Lock()
+	defer d.mu.Unlock()
```
Comment (phinze) [inline terraform/diff.go]: Why lock the mutex here instead of inside `requiresNew()`?

File: terraform/diff.go [hunk 3]
```diff
@@ -379,24 +399,103 @@ func (d *InstanceDiff) RequiresNew() bool {
...
+}
+
+// Safely copies the Attributes map
+func (d *InstanceDiff) CopyAttributes() map[string]*ResourceAttrDiff {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	attrs := make(map[string]*ResourceAttrDiff)
+	for k, v := range d.Attributes {
+		attrs[k] = v
```
Comment (phinze) [inline terraform/diff.go]: Noting that `ResourceAttrDiff` is not deep copied here. Offhand, I don't believe there's an operative difference.

**Comment by phinze** [comment]
Bless your concurrent gopher heart! This looks great. A few Qs inline.

File: config/raw_config.go (hunk 1 above)
Comment (jbardin) [inline config/raw_config.go]: Yeah, this was a "stop the failures" change. I'll check if this can return a copy.

File: terraform/diff.go (hunk 2 above)
Comment (jbardin) [inline terraform/diff.go]: `requiresNew` is called from code than already holds the lock. A pattern I often use is have an exported method which takes a mutex and delegates to an private method to do the work. This way locked methods can call one another through the private interface.

File: terraform/diff.go (hunk 3 above)
Comment (jbardin) [inline terraform/diff.go]: Nope. This is strictly to protect the map. (this was the root of the runtime panics)

File: terraform/diff.go (hunk 2 above)
Comment (phinze) [inline terraform/diff.go]: Yeah now that I've read down to the bottom I see the pattern at play. 👍

**Comment by phinze** [comment]
https://www.youtube.com/watch?v=xoHECVnQC7A

LGTM 🐎

**Pull request merged**